rfcs/007-background-workers.md

```diff
@@ -0,0 +1,299 @@
+either one of:
+
+ - Update the `JobState` to `"completed"` and leave the job in the jobs table.
+   It would be up to the job creation logic to inspect and remove it.
+
+ - Delete the job from the system altogether. This may be useful when running
+   `_replicate` jobs, for example.
+
+
+If a user decided to cancel a job that is running, they MUST toggle `CancelReq`
+job field to `true`. Then, they MUST wait for the job to stop running. If the
+job is not running, they may directly remove the job from the pending and the
+main jobs table.
+
+### Worker Capacity and Long Running Jobs
+
+Some background jobs, like continuous replication jobs, might run for years
+without interruption once they start executing. In that case, workers MAY
+choose to periodically stop some jobs and re-insert them into the pending queue
+for execution, and choose to run other jobs which have been waiting. This
+behavior is job type specific. Ensuring fairness and preventing starvation is
+also job type-specific.
+
+Worker MAY decide how often and how many jobs they pull from the pending queue.
+For example, they could use a dampening scheme where the longer their active
+queue grows, the slower the rate at which they pull jobs from the queue. That
+behavior is job type specific.
+
+Another advanced scheme is where workers could start on demand based on
+the pending queue backlog. When the backlog subsides they would stop
+(with appropriate dampening factors to reduce flapping). That is again job type
+specific, and is outside the scope of this RFC.
+
+## Worker Lifecyle
+
+Workers can be dynamically started and stopped. When workers start, they MUST
+register themselves in the `"workers"` subspace based on their `WorkerType`.
+Each change to the workers list, should be accompanied by updating the
+`"workers_vs"` versionstamp. That versionstamp MAY be efficiently monitored for
+changes using a watch.
+
+
+### Worker Liveness Monitoring
+
+As long as the workers are active and ready to process jobs, they MUST
+periodically ping their health key. The period MUST be less than the
+`HealthTimeout` value.
+
+Worker nodes MUST monitor their neighbors for liveness. When they detect an
+unresponsive neighbor, they MUST remove their jobs and re-enqueue them in the
+pending queue, then remove the dead neighbor from the workers membership list
+as well. To avoid all worker nodes, monitoring all other workers, each worker
+will monitor their neighbor to "right" in the sorted list of live workers.
+
+### Liveness and Race Conditions
+
+In case workers come back after not updating their health entry often enough,
+and another worker had cleared and re-submitted their jobs, they MUST
+monitor their own presence in the membership list, and, if they detect they
+were removed, MUST stop running all their jobs. Then they MUST re-execute their
+startup logic: add themselves to the membership list, restart health key pings,
+and start picking jobs off of the pending jobs queue, and so on.
```

Comment:
again, I'm lost. This implies that enforcing Mutual Exclusion is really the job of the workers to implement independently, in which case why have a framework?

Reply:
Good point. Update the RFC to contain a separate section regarding mutual exclusion.